[Enhancement] Convert metric to Json format

chaoyli opened the pull request:
Add a JSON format for existing metrics like this.
```
{
    "tags":
    {
        "metric":"thread_pool",
        "name":"thrift-server-pool",
        "type":"active_thread_num"
    },
    "unit":"number",
    "value":3
}
```
I add a new JsonMetricVisitor to handle the convertion.
It's not to modify existing PrometheusMetricVisitor and SimpleCoreMetricVisitor.
Also I add a unit item to indicate the metric better.

chaoyli committed:
[Feature] Convert metric to Json format

Add a JSON format for existing metrics like this.
```
{
    "tags":
    {
        "metric":"thread_pool",
        "name":"thrift-server-pool",
        "type":"active_thread_num"
    },
    "unit":"number",
    "value":3
}
```
I add a new JsonMetricVisitor to handle the convertion.
It's not to modify existing PrometheusMetricVisitor and SimpleCoreMetricVisitor.
Also I add a unit item to indicate the metric better.

chaoyli commented:
#3636

morningman commented on be/src/http/action/metrics_action.cpp:
```diff
@@ -143,15 +143,74 @@ void SimpleCoreMetricsVisitor::visit(const std::string& prefix,
+        for (auto& it : collector->metrics()) {
+            const MetricLabels& labels = it.first;
+            SimpleMetric* metric = reinterpret_cast<SimpleMetric*>(it.second);
+            _ss << "{\n\t\"tags\":\n\t{\n";
```
Why not using repidjson lib to generate the json object?

chaoyli replied:
I change it to rapidjson

morningman commented on be/src/http/action/metrics_action.cpp:
```diff
@@ -143,15 +143,74 @@ void SimpleCoreMetricsVisitor::visit(const std::string& prefix,
+        SimpleCoreMetricsVisitor visitor;
+        _metrics->collect(&visitor);
+        str.assign(visitor.to_string());
+    } else if (type == "agent") {
```
what is `agent` mean?

chaoyli replied:
The request is comes from a agent installed outof doris_be.

morningman commented on fe/src/main/java/org/apache/doris/metric/JsonMetricVisitor.java:
```diff
@@ -0,0 +1,104 @@
+ * # HELP doris_fe_job_load_broker_cost_ms doris_fe_job_load_broker_cost_ms 
+ * # TYPE doris_fe_job_load_broker_cost_ms gauge 
+ * doris_fe_job{job="load", type="mini", state="pending"} 0
+ */
```
Modify the comment

morningman commented on fe/src/main/java/org/apache/doris/metric/JsonMetricVisitor.java:
```diff
@@ -0,0 +1,104 @@
+    private static final String JVM_OLD_GC = "jvm_old_gc";
+    private static final String JVM_THREAD = "jvm_thread";
+
+    private static final String HELP = "# HELP ";
```
Useless

morningman commented on fe/src/main/java/org/apache/doris/metric/JsonMetricVisitor.java:
```diff
@@ -0,0 +1,104 @@
+        sb.append("\t\"unit\":\"").append(metric.getUnit().name().toLowerCase()).append( "\",\n");
+
+        // value
+        sb.append("\t\"value\":").append(metric.getValue().toString()).append("\n}");
```
You can use Gson lib to generate json object

chaoyli replied:
Gson can convert the object to JSON. But in our MetricLabel, key and value is all defined as variable. If I use Gson to convert to JSON, the result is not what I wanted.

morningman commented on fe/src/main/java/org/apache/doris/metric/SimpleCoreMetricVisitor.java:
```diff
@@ -56,6 +56,9 @@
     public static final String MAX_TABLET_COMPACTION_SCORE = "max_tablet_compaction_score";
 
+    private int ordinal = 0;
+    private int metric_number = 0;
```
```suggestion
    private int metricNumber = 0;
```

morningman commented on fe/src/main/java/org/apache/doris/metric/SimpleCoreMetricVisitor.java:
```diff
@@ -71,6 +74,11 @@ public SimpleCoreMetricVisitor(String prefix) {
     }
 
+    @Override
+    public void setMetricNumber(int metric_number) {
```
```suggestion
    public void setMetricNumber(int metricNumber) {
```

morningman commented on fe/src/main/java/org/apache/doris/metric/PrometheusMetricVisitor.java:
```diff
@@ -52,10 +52,18 @@
     private static final String TYPE = "# TYPE ";
 
+    private int ordinal = 0;
+    private int metric_number = 0;
```
```suggestion
    private int metricNumber = 0;
```

morningman commented on fe/src/main/java/org/apache/doris/metric/PrometheusMetricVisitor.java:
```diff
@@ -52,10 +52,18 @@
     }
 
+    @Override
+    public void setMetricNumber(int metric_number) {
```
```suggestion
    public void setMetricNumber(int metricNumber) {
```

morningman commented on fe/src/main/java/org/apache/doris/qe/StmtExecutor.java:
```diff
@@ -683,6 +684,7 @@ private void handleInsertStmt() throws Exception {
                 return;
             }
 
+            MetricRepo.COUNTER_LOAD_FINISHED.increase(1L);
```
What is the `COUNTER_LOAD_FINISHED` means? If it means that load is done successfully, 
this counter should be added after `commitAndPublishTransaction()`

morningman commented on fe/src/main/java/org/apache/doris/service/FrontendServiceImpl.java:
```diff
@@ -757,6 +759,7 @@ private boolean loadTxnCommitImpl(TLoadTxnCommitRequest request) throws UserExce
             throw new UserException("unknown database, database=" + dbName);
         }
 
+        MetricRepo.COUNTER_LOAD_FINISHED.increase(1L);
```
after `commitAndPublishTransaction()`?

morningman commented:
Why we need a json format metric output?

chaoyli commented:
> Why we need a json format metric output?

It's for incorporting another Monitor Tools easily.

chaoyli committed:
[Feature] Convert metric to Json format

Add a JSON format for existing metrics like this.
```
{
    "tags":
    {
        "metric":"thread_pool",
        "name":"thrift-server-pool",
        "type":"active_thread_num"
    },
    "unit":"number",
    "value":3
}
```
I add a new JsonMetricVisitor to handle the convertion.
It's not to modify existing PrometheusMetricVisitor and SimpleCoreMetricVisitor.
Also I add a unit item to indicate the metric better.

I also add cloning tablet into statistic divided by database.

morningman commented on fe/src/main/java/org/apache/doris/task/AgentTaskQueue.java:
```diff
@@ -215,6 +217,20 @@ public static synchronized int getTaskNum() {
         return taskNum;
     }
 
+    public static synchronized Multimap<Long, Long> getTabletIdsByDbId(TTaskType type) {
```
```suggestion
    public static synchronized Multimap<Long, Long> getTabletIdsByType(TTaskType type) {
```

morningman commented on fe/src/main/java/org/apache/doris/task/AgentTaskQueue.java:
```diff
@@ -215,6 +217,20 @@ public static synchronized int getTaskNum() {
 
+    public static synchronized Multimap<Long, Long> getTabletIdsByDbId(TTaskType type) {
+        Multimap<Long, Long> tabletIds = HashMultimap.create();
+        tabletIds.clear();
```
why call clear here?

morningman commented:
> > Why we need a json format metric output?
> 
> It's for incorporting another Monitor Tools easily. Json will be a good choice for this demand.

If we adopt this json form of expansion each new line, will it result in a very large result set?
Which monitoring system uses Json?

chaoyli commented:
> > > Why we need a json format metric output?
> > 
> > 
> > It's for incorporating another Monitor Tools easily. Json will be a good choice for this demand.
> 
> If we adopt this json form of expansion each new line, will it result in a very large result set?
> Which monitoring system uses Json?

I have test it, It will not larger than a query profile.
open-falcon opensourced by Xiaomi is use JSON to collect metric.
http://open-falcon.org/falcon-plus/

chaoyli committed:
[Feature] Change some function name to better understanding
And also use white space to replace newline in audit.log

morningman commented on fe/src/main/java/org/apache/doris/service/FrontendServiceImpl.java:
```diff
@@ -757,10 +759,15 @@ private boolean loadTxnCommitImpl(TLoadTxnCommitRequest request) throws UserExce
+                        5000, TxnCommitAttachment.fromThrift(request.txnCommitAttachment));
+        if (ret) {
+            // if commit and publish is success, load can be regarded as success
+            MetricRepo.COUNTER_LOAD_FINISHED.increase(1L);
```
public static LongCounterMetric COUNTER_TXN_REJECT;
    public static LongCounterMetric COUNTER_TXN_BEGIN;
    public static LongCounterMetric COUNTER_TXN_FAILED;
    public static LongCounterMetric COUNTER_TXN_SUCCESS;

We already have 4 metrics to monitor the transaction, is it necessary to add `COUNTER_LOAD_FINISHED` and `COUNTER_LOAD_ADD`?

chaoyli replied:
I thins COUNTER_TXN_SUCCESS will use for another logic like schema change/rollup/clone in the future. It may be confused.

morningman reviewed (commented):
LGTM, except for some little problem.
And please fix UT.

chaoyli committed:
Use rapidjson instead

chaoyli committed:
Fix unit test

chaoyli commented:
> LGTM, except for some little problem.
> And please fix UT.

I fix the UT

morningman reviewed (approved):
LGTM
